src/spreadsheets.js

```diff
@@ -0,0 +1,286 @@
+var fs = require('fs');
+var {google} = require('googleapis');
+var config = require('../config.json');
+var fluid = fluid || require('infusion');
+
+var CREDENTIALS_PATH = config.credentialsPath;
+var TOKEN_PATH = config.tokenPath;
+var SCOPES = config.scopes;
+
+/**********************
+ * SPREADSHEET GRADES *
+ **********************/
+
+var spreadsheets = fluid.registerNamespace('spreadsheets');
+module.exports = spreadsheets;
+
+// This logging function is used to trigger a breakpoint and get component tree state
+spreadsheets.debug = function(...messages) {
+    console.log(messages);
+}
+
+/**
+ * component for reading and writing entire spreadsheets, located based on their id/URL
+ */
+fluid.defaults('spreadsheets.spreadsheet', {
+    gradeNames: ['fluid.modelComponent'],
+    spreadsheetId: '', // the id of the host spreadsheet, which can be extracted from the URL
+    members: {
+        sheetBounds: {} // the coordinate ranges containing all data in each sheet 'tab', keyed by the sheetName
+        // this may be significantly larger than the used cells without incurring any computational cost
+    },
+    components: {
+        apiClient: { // the authorized API client
+            type: 'spreadsheets.sheetsAPIClient'
+        }
+    },
+    // TODO: ask A about old-style event passing versus current setup, 
+    // which is an onCreate listener triggering an invoker 
+    // '{that}.apiClient.events.onResolve': 'spreadsheets.cell.updateValue',
+    // '{that}.apiClient.events.onError': 'spreadsheets.cell.displayError'
+    listeners: { // read the remote spreadsheet on component creation
+        'onCreate.getRemoteContent': 'spreadsheets.spreadsheet.getRemoteContent({that}, {apiClient})'
+    },
+    // When the model is written to, update the remote sheet
+    // TODO: where is the change source stored?
+    modelListeners: {
+        '*': {
+            func: 'spreadsheets.spreadsheet.setRemoteContent',
+            args: ['{that}', '{apiClient}', '{change}.path.0', '{change}.value'],
```

Comment:
Looks like you are aware that the "*" style produces will produce one message per sheet, with the first path segment allocated to this, but thought I would double-check. In future we might want more fine-grained binding. Note that there is a very inefficient utility fluid.model.diff that you could make use of quickly enough https://github.com/amb26/infusion/blob/FLUID-6145/src/framework/core/js/DataBinding.js#L1584 - or go for something more heavyweight and specialised in future.